frontend/src/scenes/funnels/FunnelBarGraph.tsx

```diff
@@ -0,0 +1,140 @@
+import React, { useRef, useState } from 'react'
+import { humanizeNumber, pluralize } from 'lib/utils'
+import { FunnelStep } from '~/types'
+import { PropertyKeyInfo } from 'lib/components/PropertyKeyInfo'
+import { Button } from 'antd'
+import { ClockCircleOutlined, UserOutlined } from '@ant-design/icons'
+import { useResizeObserver } from 'lib/utils/responsiveUtils'
+import { SeriesGlyph } from 'lib/components/SeriesGlyph'
+
+import './FunnelBarGraph.scss'
+
+interface FunnelBarGraphProps {
+    layout?: 'horizontal' | 'vertical'
+    steps: FunnelStep[]
+}
+
+interface BarProps {
+    percentage: number
+    name?: string
+}
+
+type LabelPosition = 'inside' | 'outside'
+
+function Bar({ percentage, name }: BarProps): JSX.Element {
+    const barRef = useRef<HTMLDivElement | null>(null)
+    const labelRef = useRef<HTMLDivElement | null>(null)
+    const [labelPosition, setLabelPosition] = useState<LabelPosition>('inside')
+    const LABEL_POSITION_OFFSET = 8 // Defined here and in SCSS
+
+    function decideLabelPosition(): void {
+        // Place label inside or outside bar, based on whether it fits
+        const barWidth = barRef.current?.clientWidth ?? null
+        const labelWidth = labelRef.current?.clientWidth ?? null
+
+        if (barWidth !== null && labelWidth !== null) {
+            if (labelWidth + LABEL_POSITION_OFFSET * 2 > barWidth) {
+                setLabelPosition('outside')
+                return
+            }
+        }
+        setLabelPosition('inside')
+    }
+
+    useResizeObserver({
+        callback: decideLabelPosition,
+        element: barRef,
+    })
+
+    return (
+        <div className="funnel-bar-wrapper">
+            <div ref={barRef} className="funnel-bar" style={{ width: `${percentage}%` }}>
+                <div
+                    ref={labelRef}
+                    className={`funnel-bar-percentage ${labelPosition}`}
+                    title={name ? `Users who did ${name}` : undefined}
+                    aria-role="progressbar"
+                    aria-valuemin={0}
+                    aria-valuemax={100}
+                    aria-valuenow={percentage}
+                >
+                    {humanizeNumber(percentage, 2)}%
+                </div>
+            </div>
+        </div>
+    )
+}
+
+interface ValueInspectorButtonProps {
+    icon?: JSX.Element
+    onClick: (e?: React.SyntheticEvent) => any
+    children: React.ReactNode
+    disabled?: boolean
+}
+
+function ValueInspectorButton({ icon, onClick, children, disabled = false }: ValueInspectorButtonProps): JSX.Element {
+    return (
+        <Button type="link" icon={icon} onClick={onClick} className="funnel-inspect-button" disabled={disabled}>
+            <span className="funnel-inspect-label">{children}</span>
+        </Button>
+    )
+}
+
+export function FunnelBarGraph({ layout = 'horizontal', steps: stepsParam }: FunnelBarGraphProps): JSX.Element {
+    const steps = stepsParam.sort((a, b) => a.order - b.order)
```

Comment:
`sort` mutates the existing array, so in this case you're modifying the parameters passed to the funnel. Make a clone of the existing array if you wish to sort it. `[...stepsParam].sort()`

Reply:
The idea here was that I'd mutate it once at the top of the function, then never again. Will change to what you suggest, though.

Reply:
Oh no no no no, we do not run any code that has side effects inside the body of a react component.

Reply:
The point here is not that "steps" are sorted, but that "stepsParam" is sorted, wherever it was stored before it was passed to the component.